### PR TITLE
MOBT-314: Add radius_of_vicinity coordinate to vicinity processed diagnostics

### PR DESCRIPTION
**Pull request by bayliffe**
Adds a `radius_of_vicinity` coordinate to cubes processed using the `OccurrenceInVicinity` plugin.

Several changes have been made around this:
- Slight refactor of the `OccurenceInVicinity` plugin along with a new method to add the coordinate.
- Changes to the threshold plugin to ensure threshold cube slices have a `probability_of` name which is used to identify them as such.
- `_in_vicinity` name setting added within the `OccurrenceInVicinity` plugin so that the returned cubes have names reflecting the processing they have undergone.
- Removal of `_in_vicinity` name setting in the threshold CLI and vicinity CLI.

To help with review, at [commit](https://github.com/metoppv/improver/commit/bb7b111fab5493ad48aa6e76a8612a119013a60f) `bb7b111fab5493ad48aa6e76a8612a119013a60f` I have excluded the addition of the new `radius_of_vicinity` coordinate. This means that all the other changes, which do not affect the acceptance test data, can be checked at this commit by running the complete set of tests against the master test data.

The new acceptance test data includes the additional `radius_of_vicinity` coordinate, and is against this that the finished PR can be tested.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver/utilities/spatial.py [hunk 1]
```diff
@@ -490,6 +487,52 @@ def maximum_within_vicinity(self, cube: Cube) -> Cube:
...
+    def _add_vicinity_coordinate(self, cube):
+        """
+        Add a coordinate to the returned cube that records the vicinity radius
+        that has been applied to the data.
+
+        Args:
+            cube:
+                Vicinity processed cube.
+        """
+        if self.native_grid_point_radius:
```
Comment (fionaRust) [inline improver/utilities/spatial.py:519]: Do we actually need to support native grid point radius now?

File: improver/cli/vicinity.py [hunk 2]
```diff
@@ -66,10 +66,6 @@ def process(cube: cli.inputcube, vicinity: float):
...
             Cube with the vicinity processed data.
     """
-    from improver.metadata.probabilistic import in_vicinity_name_format
     from improver.utilities.spatial import OccurrenceWithinVicinity
 
-    result = OccurrenceWithinVicinity(radius=vicinity).process(cube)
-    result.rename(in_vicinity_name_format(result.name()))
-
-    return result
+    return OccurrenceWithinVicinity(radius=vicinity).process(cube)
```
Comment (fionaRust) [inline improver/cli/vicinity.py:71]: The acceptance test data for this CLI needs updating too?

File: improver/threshold.py [hunk 3]
```diff
@@ -342,6 +357,7 @@ def process(self, input_cube: Cube) -> Cube:
 
             self._add_threshold_coord(cube, threshold)
             cube.coord(var_name="threshold").convert_units(input_cube.units)
+            self._update_metadata(cube, update_cell_methods=False)
```
Comment (fionaRust) [inline improver/threshold.py:360]: Do you need to update the cube name twice, or could you split this function up to make it clearer what's being updated each time you call it?

**Review by fionaRust — CHANGES_REQUESTED** [review]

File: improver/utilities/spatial.py (hunk 1 above)
Comment (bayliffe) [reply improver/utilities/spatial.py:519]: Yes, this was added for precip phase change to provide a non-zero neighbourhood size when calculating maximum orography in neighbourhood. This is still being used, e.g. [here](https://github.com/MetOffice/improver_suite/blob/2382e71199b10a8bb5987e74e62d7b86060e31f1/lib/python/mo/settings/app/phase_change_level.py#L25). We've added another step after that to fix up any remaining NaNs, but the neighbourhood approach is preferable to the applying only the latter.

File: improver/cli/vicinity.py (hunk 2 above)
Comment (bayliffe) [reply improver/cli/vicinity.py:71]: Yep, too many overlapping PRs and acceptance test data sets for me to keep straight apparently. Done.

File: improver/threshold.py (hunk 3 above)
Comment (bayliffe) [reply improver/threshold.py:360]: I need to check I understand what you are after.

Are you suggesting I split out the setting of the name, cell method, and the rest (grouping cube units, coordinate name and attributes together) into three separate functions and call each one only where required?

I would then need to call all three in sequence in `LatitudeDependentThreshold` which also uses this function, or I could write a wrapper for cases where we want to call them all (but that's not so different from where it is right now).

File: improver/utilities/spatial.py (hunk 1 above)
Comment (fionaRust) [reply improver/utilities/spatial.py:519]: Ok fair enough

File: improver_tests/utilities/test_OccurrenceWithinVicinity.py [hunk 4]
```diff
@@ -120,35 +140,58 @@ def test_basic(cube, binary_expected, kwargs):
...
-    """Test that if a zero radius / grid point radius, or no radius / grid point
-    radius is provided, the input data is returned unchanged. This test uses
-    both an equal area and latlon grid as a 0 radius can be applied to both."""
+def test_zero_radius(request, fixture_name, keyword):
+    """Test that if a zero radius / grid point radius is provided, the input
+    data is returned unchanged. This test uses both an equal area and latlon
+    grid as a 0 radius can be applied to both."""
     cube = request.getfixturevalue(fixture_name)
-    kwargs = {keyword: value}
+    kwargs = {keyword: 0}
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/utilities/test_OccurrenceWithinVicinity.py]: I've changed the behaviour to raise an exception in cases where both radius and grid_point_radius are 0. If one radius is set to a non-zero value, that radius will be used.

**Review by fionaRust — COMMENTED** [review]
Tiny docstring change needed

File: improver/threshold.py (hunk 3 above)
Comment (fionaRust) [reply improver/threshold.py]: This is no longer relevant

File: improver/cli/vicinity.py (hunk 2 above)
Comment (fionaRust) [reply improver/cli/vicinity.py:71]: thanks

File: improver_tests/utilities/test_OccurrenceWithinVicinity.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fionaRust) [reply improver_tests/utilities/test_OccurrenceWithinVicinity.py]: thanks

**Review by fionaRust — APPROVED** [review]

**Review by MoseleyS — CHANGES_REQUESTED** [review]
A few suggestions for you

File: improver/utilities/spatial.py [hunk 5]
```diff
@@ -433,7 +448,7 @@ def __init__(
             self.land_mask = None
         self.land_mask_cube = land_mask_cube
 
-    def maximum_within_vicinity(self, cube: Cube) -> Cube:
+    def maximum_within_vicinity(self, cube: Cube, grid_point_radius) -> Cube:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/spatial.py:451]: Done.

File: improver_tests/utilities/test_OccurrenceWithinVicinity.py [hunk 6]
```diff
@@ -120,38 +140,48 @@ def test_basic(cube, binary_expected, kwargs):
...
     assert isinstance(result, Cube)
     assert np.allclose(result.data, binary_expected)
 
 
+def test_metadata(cube, radius_coord, grid_point_radius_coord):
+    """Test that the metadata on the cube reflects the data it contains
+    following the application of vicinity processing."""
+    # defining the radius
+    plugin = OccurrenceWithinVicinity(radius=RADIUS)
+    # repeat thet test with the same plugin instance to ensure self variables
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/utilities/test_OccurrenceWithinVicinity.py:153]: Done.

File: improver_tests/utilities/test_OccurrenceWithinVicinity.py [hunk 7]
```diff
@@ -120,38 +140,48 @@ def test_basic(cube, binary_expected, kwargs):
...
+    # have not been modified.
+    for test in range(2):
+        result = plugin.process(cube)
+        assert isinstance(result, Cube)
+        assert result.coord("radius_of_vicinity") == radius_coord
+        assert "in_vicinity" in result.name()
+
+    # defining the grid point radius
+    plugin = OccurrenceWithinVicinity(grid_point_radius=GRID_POINT_RADIUS)
+    # repeat thet test with the same plugin instance to ensure self variables
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/utilities/test_OccurrenceWithinVicinity.py:163]: Done.

File: improver_tests/utilities/test_OccurrenceWithinVicinity.py [hunk 8]
```diff
@@ -120,38 +140,48 @@ def test_basic(cube, binary_expected, kwargs):
...
+    # have not been modified.
+    for test in range(2):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/utilities/test_OccurrenceWithinVicinity.py:165]: Done.

File: improver_tests/utilities/test_OccurrenceWithinVicinity.py [hunk 9]
```diff
@@ -120,38 +140,48 @@ def test_basic(cube, binary_expected, kwargs):
...
+    plugin = OccurrenceWithinVicinity(radius=RADIUS)
+    # repeat thet test with the same plugin instance to ensure self variables
+    # have not been modified.
+    for test in range(2):
+        result = plugin.process(cube)
+        assert isinstance(result, Cube)
+        assert result.coord("radius_of_vicinity") == radius_coord
+        assert "in_vicinity" in result.name()
+
+    # defining the grid point radius
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/utilities/test_OccurrenceWithinVicinity.py:161]: Done.

File: improver_tests/utilities/test_OccurrenceWithinVicinity.py [hunk 10]
```diff
@@ -444,11 +474,60 @@ def test_no_realization_or_time(request, cube_with_realizations, land_fixture):
...
+    expected = (
+        "Vicinity processing requires that one of radius or "
+        "grid_point_radius should be set to a non-zero value"
+    )
+
+    with pytest.raises(ValueError, match=expected):
+        OccurrenceWithinVicinity(radius=radius, grid_point_radius=grid_point_radius)
+
+
+def test_set_grid_point_radius(cube):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/utilities/test_OccurrenceWithinVicinity.py:506]: Done.

File: improver/utilities/spatial.py [hunk 11]
```diff
@@ -417,9 +418,23 @@ def __init__(
...
                 like mask value.
+
+        Raises:
+            ValueError: If both radius and grid point radius are set.
+            ValueError: If neither radius or grid point radius are set.
+            ValueError: Land mask not named land_binary_mask.
         """
-        if radius is not None and grid_point_radius is not None:
-            raise ValueError("Only one of radius or grid_point_radius should be set")
+        if radius and grid_point_radius:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/utilities/spatial.py:427]: Added.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Spotted a few more minor tweaks.

**Review by MoseleyS — APPROVED** [review]

**Pull request merged**